### PR TITLE
Implement porcelain grep

### PR DESCRIPTION
**Pull request by betesh**

### CODE REVIEW

**Comment by betesh** [comment]
@bartkamphorst Waiting for your feedback...

File: lib/rjgit.rb [hunk 1]
```diff
@@ -190,6 +190,40 @@ def self.describe(repository, ref, options = {})
...
+    #  * ref
+    #  * path_filter
+    #  * case_insensitive
+    def self.grep(repository, query, options={})
+      case_insensitive = options[:case_insensitive]
+      repo = RJGit.repository_type(repository)
+      walk = RevWalk.new(repo)
+      ls_tree_options = {:recursive => true, :path_filter => options[:path_filter]}
+
+      raise "A #{query.class} was passed to #{self}.grep().  Only Regexps and Strings are supported!" unless String === query || Regexp === query
```
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/rjgit.rb:204]: Also, the method is now effectively evaluating `query` for `String` or `Regexp` three times, which seems redundant. What about, when a String is given, we turn it into a Regexp and subsequently work with that? Then we can just use `rows.grep(query)`.

[the remark this answers is not in the document]
Comment (dometto) [reply lib/rjgit.rb:204]: Alternatively, can't we just add an `else` clause to the `case` statement below and perform the `raise` there?

File: lib/rjgit.rb [hunk 2]
```diff
@@ -190,6 +190,40 @@ def self.describe(repository, ref, options = {})
...
+
+      if case_insensitive
+        query = case query
+        when Regexp then Regexp.new(query.source, query.options | Regexp::IGNORECASE)
+        when String then query.downcase
+        end
```
Comment (dometto) [inline lib/rjgit.rb:210]: I meant add an `else` clause here. Raising an exception should still work even though the `case` statement is used in an assignment:

```ruby
2.6.3 :001 > query = raise "FAIL"
Traceback (most recent call last):
[...]
RuntimeError (FAIL)
```

File: lib/rjgit.rb [hunk 3]
```diff
@@ -190,6 +190,40 @@ def self.describe(repository, ref, options = {})
...
+      end
+
+      ls_tree(repo, nil, options.fetch(:ref, 'HEAD'), ls_tree_options).each_with_object({}) do |item, result|
+        blob = Blob.new(repo, item[:mode], item[:path], walk.lookup_blob(ObjectId.from_string(item[:id])))
+        next if blob.binary?
+
+        rows = blob.data.split("\n")
+        data = case query
+        when Regexp then rows.grep(query)
```
Comment (dometto) [inline lib/rjgit.rb:219]: Ah, didn't see the second `case` statement here. On board with @bartkamphorst's suggestion of converting string to regex. I think you can use `Regex.new` or `%r{#{string}}`. See [here](https://stackoverflow.com/questions/8652715/convert-a-string-to-regular-expression-ruby).

File: lib/rjgit.rb [hunk 4]
```diff
@@ -190,6 +190,40 @@ def self.describe(repository, ref, options = {})
...
+        when Regexp then Regexp.new(query.source, query.options | Regexp::IGNORECASE)
+        when String then query.downcase
+        end
+      end
+
+      ls_tree(repo, nil, options.fetch(:ref, 'HEAD'), ls_tree_options).each_with_object({}) do |item, result|
+        blob = Blob.new(repo, item[:mode], item[:path], walk.lookup_blob(ObjectId.from_string(item[:id])))
+        next if blob.binary?
+
+        rows = blob.data.split("\n")
```
Comment (dometto) [inline lib/rjgit.rb:217]: What about adding an `&block` optional parameter to this method, and `yield blob.data` or `yield rows` here if `block_given?`?

File: lib/rjgit.rb [hunk 5]
```diff
@@ -190,6 +190,40 @@ def self.describe(repository, ref, options = {})
...
+
+    # options:
+    #  * ref
+    #  * path_filter
+    #  * case_insensitive
+    def self.grep(repository, query, options={})
+      case_insensitive = options[:case_insensitive]
+      repo = RJGit.repository_type(repository)
+      walk = RevWalk.new(repo)
+      ls_tree_options = {:recursive => true, :path_filter => options[:path_filter]}
```
[the remark this answers is not in the document]
Comment (betesh) [reply lib/rjgit.rb:202]: > Maybe let :recursive also be set in the options hash?

It's hard to imagine a case where you wouldn't want this to be recursive.  If you did, you could use path_filter.

> Also, stylistically I have a slight preference for the more recent hash syntax

I was following the other examples in this file.  According to the gemspec, we don't require any minimum version of ruby.  If I use JSON Hash Syntax, we'd have to add `required_ruby_version` to the gemspec.  What's the minimum version you want to support?

File: lib/rjgit.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (betesh) [reply lib/rjgit.rb:204]: If there were only one case clause, I'd have `raise` in an `else` there.  Given there are 2 cases clauses, putting it only in the first looks inconsistent and sloppy, but putting it in both is redundant.

I like the idea of converting String to Regexp.  I'm going to make that change.

File: lib/rjgit.rb (hunk 4 above)
Comment (betesh) [reply lib/rjgit.rb:217]: Can you show me an example of a block you might pass to `Gollum::Git.grep`?  I don't see any uses of grep in rjgit_adapter itself and I'm not sure what you have in mind.  It seems that that would require the user to parse the blob data.  Personally I'd rather just pass a Regex to `grep` and let the library figure out how to parse it.

File: lib/rjgit.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (betesh) [reply lib/rjgit.rb]: Done!

File: lib/rjgit.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/rjgit.rb:202]: Good point. Leave it as is; I'll refactor once we sort out which minimum version we support.

File: lib/rjgit.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/rjgit.rb]: 👍 Thanks!

**Comment by betesh** [comment]
@bartkamphorst I'm pretty sure I addressed all your concerns.  @dometto I'm not sure what the purpose of `yield` would be--still need to see an example.

**Comment by bartkamphorst** [comment]
> I'm pretty sure I addressed all your concerns.

Agreed, thanks. @dometto ?

**Comment by dometto** [comment]
>  @dometto I'm not sure what the purpose of yield would be--still need to see an example.

The use case I had in mind is pretty thin: gollum needs it. So I thought that if these function supported a block, I could simplify the implementation [here](https://github.com/gollum/rjgit_adapter/blob/master/lib/rjgit_adapter/git_layer_rjgit.rb#L179). I imagined there might be other uses for it, but you maybe right it's too specfic.

But the needs of another project shouldn't hold up this PR. Better to merge this, and refactor RJGit and/or the adapter at a later point.

**Comment by betesh** [comment]
@dometto Can you show me where `grep` is _called_ (not defined) with a block?  I just don't understand what sort of code would go in this block.

**Comment by dometto** [comment]
See https://github.com/gollum/gollum-lib/blob/master/lib/gollum-lib/wiki.rb#L495

The idea is: the grep method yields `filename, data` pairs, the caller defines how to 'keep score' of the hits in the provided block.

But again, no real need to ponder this now!

**Comment by betesh** [comment]
@bartkamphorst are you ready to merge?

**Comment by bartkamphorst** [comment]
Thanks @betesh for your contributions, we appreciate it!

**Comment by betesh** [comment]
Thanks!  Can you release this?

**Comment by bartkamphorst** [comment]
> Thanks! Can you release this?

Done! v. 5.7.0.2 is out.

**Comment by betesh** [comment]
Thanks!

**Pull request merged**